### PR TITLE
Chap 2 done

### PR DESCRIPTION
**Pull request by Pythonen**
### Solutions for Chapter {NUMBER}

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +337,7 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList a b xs = drop a (take b $ xs)
```
Comment (chshersh) [inline src/Chapter2.hs:340]: It looks like your implementation doesn't always return empty lists on negative numbers. Probably, you can think about various corner cases here :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -500,7 +502,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: [Int] -> Bool
+isThird42 [] = False
+isThird42 xs = if head(drop 2 xs) == 42 then True else False
```
Comment (chshersh) [inline src/Chapter2.hs:507]: `head` is a partial function, so if you give a list of size 1 or 2 to this function, it fails in runtime. Here you are supposed to use pattern-matching only, without any standard functions :slightly_smiling_face:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -727,7 +735,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat(map(\x -> replicate x x) l)
```
Comment (chshersh) [inline src/Chapter2.hs:738]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 4]
```diff
@@ -740,7 +748,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains x xs = filter (elem x) xs
```
Comment (chshersh) [inline src/Chapter2.hs:752]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 5]
```diff
@@ -857,7 +869,8 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind s = foldr (\x xs -> xs ++ [x]) [] s
```
Comment (chshersh) [inline src/Chapter2.hs:873]: Clever usage of standard functions! Your solution is correct. However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — COMMENTED** [review]
I left some comments about solutions in my review, but since the PR is already merged, you can only incorporate my suggestions in some other commit :octocat:

**Comment by Pythonen** [comment]
Hey and thanks for the effort of correcting me! I'm quite new to haskell, I've only done [haskell mooc](haskell.mooc.fi) as a part of my Uni studies. Interesting language and very well constructed exercises!

**Pull request merged**